compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -71,10 +68,154 @@ getSwizzle(IREE::GPU::MMAIntrinsic mma, int operandIdx) {
+// dimension corresponding to the given source dimension index.
+static int64_t
+getExpandedDimFirstIdx(const SmallVector<SmallVector<int64_t>> &expandShape,
+                       int64_t srcIndex) {
+  int dstIndexFirst = 0;
+  for (int i = 0; i < srcIndex; ++i) {
+    dstIndexFirst += expandShape[i].size();
+  }
+  return dstIndexFirst;
+}
+
+// Unroll the dimension given by `srcIndex` by the given `unrollFactor`.
+// This is not interleaving layouts. The layout will consist of multiple copies
+// of the input tile, side by side.
+//
+// Example:
+//    Input swizzle = { expandShape = [[16], [4]], permutation = [1, 0] }
+//    Input srcIndex = 1
+//    Input unrollFactor = 4
+// -> Output swizzle = { expandShape = [[16], [4, 4]], permutation = [1, 2, 0] }
+//
+static void unroll(MaterializeEncodingInfo::Swizzle &swizzle, int srcIndex,
+                   int unrollFactor) {
+  assert(unrollFactor > 1);
+  int dstIndexFirst = getExpandedDimFirstIdx(swizzle.expandShape, srcIndex);
+
+  // The new unrolling dimension is inserted at the start of the expandShape
+  // dimensions group corresponding to srcIndex.
+  swizzle.expandShape[srcIndex].insert(swizzle.expandShape[srcIndex].begin(),
+                                       unrollFactor);
+  // Since we are not interleaving here, generating side-by-side copies of the
+  // original layout, the new unrolling dimension is the new outermost
+  // dimension. Existing entries get shifted to make room for it.
+  for (auto &p : swizzle.permutation) {
+    p += (p >= dstIndexFirst);
+  }
+  swizzle.permutation.insert(swizzle.permutation.begin(), dstIndexFirst);
+}
+
+// Returns the index of the dimension whose flattened size (flattening inner
+// dimensions into it) matches the given `targetSize`.
+static int64_t getDimIdxForTargetSize(const SmallVector<int64_t> &shape,
+                                      int64_t targetSize) {
+  int interleaveAt = 0;
+  int size = shape[0];
+  for (interleaveAt = shape.size() - 1; interleaveAt >= 0; --interleaveAt) {
+    assert(size <= targetSize);
+    assert((targetSize % size) == 0);
+    if (size == targetSize) {
+      break;
+    }
+    size *= shape[interleaveAt];
+  }
+  return interleaveAt;
+}
+
+// Interleave the layout in `swizzle` by mutating `swizzle.permutation` to
+// move permutation[0], the outer-most dimension (which the unroll() function
+// created to be the unrolling dimension), to the appropriate inner dimension
+// index to ensure that the layout is interleaved exactly so that
+// `targetInterleavedElements` are made contiguous in the new layout.
+//
+// Example:
+//    Input swizzle = { expandShape = [[16], [4, 4]], permutation = [1, 2, 0] }
+//    Input srcIndex = 1
+//    Input targetInterleavedElements = 4
+// -> Output swizzle = { expandShape = [[16], [4, 4]], permutation = [2, 0, 1] }
+//
+static void interleave(MaterializeEncodingInfo::Swizzle &swizzle, int srcIndex,
+                       int targetInterleavedElements) {
```

Comment:
Thinking about this more, maybe it would be better to explicitly pass the dimension that you want to interleave at. Then there is no need for the logic that finds the index to interleave at based on load sizes. The dimension can be found from the rank of `swizzle.expandShape[srcIndex]` before unrolling, and then passed to the `interleave` function. What do you think?

Reply:
Loved this idea, thanks! So I did it; but then I realized that the caller nows needs to perform the interleaving-index computation themselves, which is actually nice, it's less magical and I believe it self-documents what `getDimIdxForTargetSize` is used for. WDYT?